### PR TITLE
Improve State Navigation using Left/Right Arrow Buttons

### LINKED ISSUE #166: Navigation between states doesn't work at high zoom level.
Pressing Left/Right arrow buttons doesn't select the prev/next available state at high zoom level. It takes the cursor directly to the start/end time range. 
Improve Navigation

### PR DESCRIPTION
**Pull request by muddana-satish**
- Added next and prev possible state fields to each row in tsp-data-provider.ts and stored the values from first and last null states.
- Start time of first null state gives the timestamp of previous possible state.
- End time of last null state gives the timestamp of next possible state.

fixes #166 

This PR is dependent on https://github.com/theia-ide/timeline-chart/pull/58

Signed-off-by: muddana-satish <satish.muddana@ericsson.com>

### CODE REVIEW

**Comment by muddana-satish** [comment]
Please include below PRs if you are testing these changes.

https://github.com/theia-ide/theia-trace-extension/pull/150
https://github.com/theia-ide/theia-trace-extension/pull/136

**Comment by muddana-satish** [comment]
**Known Issues**

Sometimes the cursor doesn't move the to the exact timestamp of the next state's start. But moves a bit more or bit less. 
I believe this is because of the loss of precision in the timestamps discussed in #41 Once we have a fix for that, this one should work fine.

Kapture 2020-10-16 at 08 43 31

Sometimes pressing Right/Left doesn't appear like the cursor is moving between start & end of very narrow states. This is because we are setting the min width of a state to 1 (for visibility) in https://github.com/theia-ide/timeline-chart/pull/40. But the actual start and end timestamps of these states is still the same(very very narrow) and it feels like the cursor didn't move. I believe this is okay? 

In the screen cast, it looks like the cursor jumps to the next state without properly navigating to start and end of each state.

Kapture 2020-10-16 at 08 58 18

**Review by tahini — COMMENTED** [review]
I'm using the timeline-chart patch, but for some reason, the selection is lost as I move right or left towards the next state. See the screenshot below, the row is deselected. Does it require one of your current patches for keeping row selection too?

left-right

**Comment by muddana-satish** [comment]
> I'm using the timeline-chart patch, but for some reason, the selection is lost as I move right or left towards the next state. See the screenshot below, the row is deselected. Does it require one of your current patches for keeping row selection too?
> 
> left-right

Yes, this requires #136. It isn't merged yet. So, we need to include those changes as well.

File: viewer-prototype/src/browser/trace-viewer/components/data-providers/tsp-data-provider.ts [hunk 1]
```diff
@@ -89,16 +80,18 @@ export class TspDataProvider {
...
         });
 
         this.timeGraphRows = newTimeGraphRows;
     }
 
-    protected getStateModelByRow(row: TimeGraphRow, chartStart: number): TimelineChart.TimeGraphRowElementModel[] {
+    private getRowModel(row: TimeGraphRow, chartStart: number, rowId: number, entry: TimeGraphEntry) {
         const states: TimelineChart.TimeGraphRowElementModel[] = [];
+        let prevPossibleState = entry.startTime;
+        let nextPossibleState = entry.endTime;
```
[the remark this answers is not in the document]
Comment (muddana-satish) [reply viewer-prototype/src/browser/trace-viewer/components/data-providers/tsp-data-provider.ts:94]: Usually typescript throws such errors. I checked the 'startTime' and 'endTime' fields in entry class are strict numbers.

**Comment by ankusht-work** [comment]
I see two issues:
1)  After navigating through states (in each direction), suddenly it stops navigating any further. Once you reset the focus or zoom in/out, it will start to navigate again.
2) Horizontal scrollbar disappears if the last state on any given row is at the very end of the timeline chart. This is not related and needs to be a separate issue.

Anyways, I believe none of the issues mentioned above should be a blocker.

**Comment by muddana-satish** [comment]
> I see two issues:
> 
> 1. After navigating through states (in each direction), suddenly it stops navigating any further. Once you reset the focus or zoom in/out, it will start to navigate again.
> 2. Horizontal scrollbar disappears if the last state on any given row is at the very end of the timeline chart. This is not related and needs to be a separate issue.
> 
> Anyways, I believe none of the issues mentioned above should be a blocker.

Thanks @ankusht-work for testing the feature. 

I believe the root cause for the first issue is same as #https://github.com/theia-ide/timeline-chart/issues/62. I will try to fix it as part of a different PR.

The second one is the same issue reported by @tahini in https://github.com/theia-ide/timeline-chart/pull/46. I created an issue https://github.com/theia-ide/timeline-chart/issues/63 for this bug.

Thanks for reporting them.

**Comment by muddana-satish** [comment]
> I'm using the timeline-chart patch, but for some reason, the selection is lost as I move right or left towards the next state. See the screenshot below, the row is deselected. Does it require one of your current patches for keeping row selection too?

@tahini I rebased the branch after meerging #136 . Can you please verify it now.

File: viewer-prototype/src/browser/trace-viewer/components/data-providers/tsp-data-provider.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (tahini) [reply viewer-prototype/src/browser/trace-viewer/components/data-providers/tsp-data-provider.ts:94]: Indeed, the code says it's mandatory. They shouldn't be, but this code doesn't know that ;-) We'll fix if/when we fix it in the corresponding upstreams.

File: viewer-prototype/src/browser/trace-viewer/components/data-providers/tsp-data-provider.ts [hunk 2]
```diff
@@ -121,6 +114,7 @@ export class TspDataProvider {
                 if (nextState && nextState.value < 0) {
                     // Add gap state
                     states.push({
+                        // We should probably remove id from state. We don't use it anywhere.
```
Comment (tahini) [inline viewer-prototype/src/browser/trace-viewer/components/data-providers/tsp-data-provider.ts:117]: Add a TODO here for that

**Review by tahini — COMMENTED** [review]
Works well now! Will need to be updated when timeline-chart patch is in, so you can add the TODO comment then too.

**Comment by bhufmann** [comment]
> I see two issues:
> 
>     1. After navigating through states (in each direction), suddenly it stops navigating any further. Once you reset the focus or zoom in/out, it will start to navigate again.
> 
BTW, this issues seems to be fixed with the latest update of the timeline-chart ([timeline-chart/58](https://github.com/theia-ide/timeline-chart/pull/58)) that this PR depends on.

**Review by bhufmann — APPROVED** [review]
Looks good to me. Thanks!

**Comment by muddana-satish** [comment]
Rebased after #147 is merged. I tested the changes and it works fine.

@tahini Could you please review it again.

**Review by tahini — APPROVED** [review]
Works well, thanks!

**Pull request merged**